include/swift/Basic/DiagnosticCollector.h

```diff
@@ -0,0 +1,80 @@
+//===--- DiagnosticCollector.h --------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This defines a diagnostic collector, collecting information required for
+// emitting, de-duplicating, and coalescing individual diagnostics into
+// something more meaningful.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_DIAGNOSTICCOLLECTOR_H
+#define SWIFT_BASIC_DIAGNOSTICCOLLECTOR_H
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SmallVector.h"
+
+#include <type_traits>
+#include <vector>
+
+namespace swift {
+
+/// A container for collecting multiple diagnostics that have a common anchor.
+template <typename AnchorType, typename DiagnosticInformationType>
```

Comment:
Why not just use `ASTNode` and parametrize only one the payload?

Reply:
Because it might not always just be an anchor location.
I'm considering using a pair in the TypeCheckEffects using the expression and the effect kind so I can pull out the 'await' diagnostics separately from the 'try' effects once I get around to fixing the bugs in the try fix-it location.

I could just as easily record this info in the diagnostic info, but then I'd have to walk all the diagnostics on the anchor instead of just seeing if the list is empty.

Reply:
I'm wondering whether `ConstraintLocator *` should somehow be fitted into `ASTNode` concept so it could represent all kinds of situations without having to allocate a locator just to store an `Expr *`. It might be simpler to just use `ASTNode` for now since it only covers effect diagnostics and take some more time to think how to unify constraint system storage with this...

Reply:
Hmm, maybe I'm not quite following. I don't see a technical reason why an `ASTNode` couldn't represent a `ConstraintLocator` in the pointer union, but from an architectural standpoint, it gets a little unpleasant. We'd have to move the definition of `ConstraintLocator` up from the Sema library into the AST library. The other part is, is a ConstraintLocator always representable by a given ASTNode? I suppose that would be the anchor? I don't know, it doesn't feel like a constraint locator "is an" ASTNode. I'm a little annoyed that the ASTNode operates the way it does instead of using inheritance because my first point wouldn't be a problem if it did use inheritance.

Though this feels a bit orthogonal to whether we should have the anchor-type be templated instead of just being an ASTNode. I don't feel super strongly, but I also don't see a reason to have that restriction arbitrarily. Is there something that I'm missing that this will preclude us from moving the allocation constraint locators?

Reply:
It's not a technical problem but a conceptual one, for me the question is either it's even worth trying to integrate collector into a constraint system at all and what grouping criterion is going to be (would it use a simplified locator or original one, does it need to account for a kind of a fix etc.). It seems that coalescing algorithm could just accept already pre-grouped diagnostics or introduce a common "location" abstraction for both type-checker effects and constraint system to use. Trying to generalize this from the start seems premature and could be done incrementally.